### PR TITLE
Initial version of reorganization

### PR DESCRIPTION
**Pull request by ayobuenavista**

### CODE REVIEW

File: .prettierignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# zeppelin contracts
+contractsSol5/utils/zeppelin
+contractsSol5/bridges
```
Comment (ilanDoron) [inline .prettierignore]: do we have bridges folder in utils project?

File: .solhint.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  "plugins": [],
+  "rules": {
+    "avoid-call-value": "error",
+    "avoid-low-level-calls": "off",
+    "avoid-sha3": "error",
+    "avoid-suicide": "error",
+    "avoid-throw": "error",
+    "avoid-tx-origin": "off",
+    "check-send-result": "error",
+    "compiler-version": ["error", "0.6.6"],
```
Comment (ilanDoron) [inline .solhint.json]: lets remove compile version.
we have multiple compile versions.

File: .solhintignore [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+## mock
+contracts/sol6/mock/TestTokenNotReturn.sol
+contracts/sol6/mock/Token.sol
+
+## utils
+contracts/sol6/utils/Utils5.sol
+contracts/sol6/utils/zeppelin/Address.sol
+contracts/sol6/utils/zeppelin/SafeERC20.sol
+contracts/sol6/utils/zeppelin/SafeMath.sol
```
Comment (ilanDoron) [inline .solhintignore]: please see if after removing compiler version.
any of the files doesn't issue errors.

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  - 10
+
+before_install:
+  - npm i -g npm
+  - npm i -g mocha-only-detector
+  - find test/ -name "*.js" -print0 | xargs -0 mocha-only-detector
+
+env:
+  global:
+    - COVERAGE_BRANCH=master
```
Comment (ilanDoron) [inline .travis.yml]: I think we can remove coverage from travis.

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+jobs:
+  include:
+    - env:
+        - TEST_PART: Sol6
+    - env:
+        - TEST_PART: Coverage
+
+install:
+  - npm ci
```
Comment (ilanDoron) [inline .travis.yml]: yarn

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+before_script:
+  - npx solhint "contracts/sol6/**/*.sol"
```
Comment (ilanDoron) [inline .travis.yml]: can add script "lint" to yarn.
and here use yarn lint.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,33 @@
...
+For more details, please visit our [developer portal](https://developer.kyber.network/)
+
+## Package Manager
+We use `yarn` as the package manager. You may use `npm` and `npx` instead, but commands in bash scripts may have to be changed accordingly.
+
+## Setup
+1. Clone this repo
+2. `yarn`
+
+## Compilation with Buidler
```
Comment (ilanDoron) [inline README.md]: please compare to my readme

I think its enough to have title "Compilation."

File: .solhintignore [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+## utils
+contracts/sol5/utils/PermissionGroups2.sol
+contracts/sol5/utils/Utils4.sol
+contracts/sol5/utils/Withdrawable2.sol
+
+## SOLIDITY 6 ##
+
+## mock
+contracts/sol6/mock/TestTokenNotReturn.sol
```
Comment (ilanDoron) [inline .solhintignore]: seems this contract is not used anywhere.

File: buidler.config.js [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+  paths: {
+    sources: "./contracts/sol6",
+    tests: "./test/sol6",
+  },
+
+  mocha: {
+    enableTimeouts: false
+  }
+};
```
Comment (ilanDoron) [inline buidler.config.js]: add line

File: contracts/sol4/mock/TokenNoDecimals.sol [hunk 10]
```diff
@@ -0,0 +1,197 @@
+pragma solidity ^0.4.18;
+
```
Comment (ilanDoron) [inline contracts/sol4/mock/TokenNoDecimals.sol:2]: seems this contract isn't used.

File: contracts/sol6/mock/ReentrancyAttack.sol [hunk 11]
```diff
@@ -0,0 +1,9 @@
+pragma solidity 0.6.6;
```
Comment (ilanDoron) [inline contracts/sol6/mock/ReentrancyAttack.sol:1]: Seems this contract isn't used here.
can remove.

File: package.json [hunk 12]
```diff
@@ -0,0 +1,58 @@
+{
+  "name": "kyber-utils-sc",
+  "version": "1.0.0",
```
Comment (ilanDoron) [inline package.json]: we decided 2.0.0

File: test/sol4/permissionGroups.js [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+
+        try {
+            await permissionsInst.addAlerter(thousandAddress);
+            assert(false, "throw was expected in line above.")
+        }
+        catch(e){
+            assert(Helper.isRevertErrorMessage(e), "expected throw but got: " + e);
+        }
+    });
+});
```
Comment (ilanDoron) [inline test/sol4/permissionGroups.js]: please add line

File: test/sol5/utils4.js [hunk 14]
```diff
@@ -0,0 +1,378 @@
+const MockUtils4 = artifacts.require("MockUtils4.sol")
+const Helper = require("../helper.js");
+const TestToken = artifacts.require("Token.sol");
```
Comment (ilanDoron) [inline test/sol5/utils4.js]: please see if TestToken usedanyhere.

File: .travis.yml (hunk 4 above)
Comment (ilanDoron) [reply .travis.yml]: lets discuss in TMR meeting and finalize

**Review by ilanDoron — COMMENTED** [review]
Looks good
note u have to update travis script to use yarn
since package-lock is missing
need to make sure any unused contracts are removed.

File: .prettierignore (hunk 1 above)
Comment (ayobuenavista) [reply .prettierignore]: Removed.

File: .solhintignore (hunk 3 above)
Comment (ayobuenavista) [reply .solhintignore]: Still gives out error, so this is retained.

File: .travis.yml (hunk 4 above)
Comment (ayobuenavista) [reply .travis.yml]: Line removed.

File: .travis.yml (hunk 5 above)
Comment (ayobuenavista) [reply .travis.yml]: Fixed.

File: .solhintignore (hunk 8 above)
Comment (ayobuenavista) [reply .solhintignore]: Removed.

File: contracts/sol4/mock/TokenNoDecimals.sol (hunk 10 above)
Comment (ayobuenavista) [reply contracts/sol4/mock/TokenNoDecimals.sol:2]: This is still used in the sol4 tests.

File: contracts/sol6/mock/ReentrancyAttack.sol (hunk 11 above)
Comment (ayobuenavista) [reply contracts/sol6/mock/ReentrancyAttack.sol:1]: Still used in the tests somehow.

File: package.json (hunk 12 above)
Comment (ayobuenavista) [reply package.json]: Fixed.

File: test/sol5/utils4.js (hunk 14 above)
Comment (ayobuenavista) [reply test/sol5/utils4.js]: Still used.

File: .travis.yml (hunk 6 above)
Comment (ayobuenavista) [reply .travis.yml]: Fixed.

File: README.md (hunk 7 above)
Comment (ayobuenavista) [reply README.md]: Changed.

File: .solhint.json [hunk 15]
```diff
@@ -9,10 +9,9 @@
     "avoid-throw": "error",
     "avoid-tx-origin": "off",
     "check-send-result": "error",
-    "compiler-version": ["error", "0.6.6"],
+    "compiler-version": "off",
     "const-name-snakecase": "error",
     "func-name-mixedcase": "error",
-    "func-order": "error",
```
Comment (ilanDoron) [inline .solhint.json]: is this removed?

File: package.json [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+  "main": "truffle.js",
+  "directories": {
+    "test": "test"
+  },
+  "scripts": {
+    "test": "./test.sh",
+    "compile": "./compile.sh",
+    "compileSol6": "./compileSol6.sh",
+    "coverage": "./coverage.sh",
+    "lint": "npx solhint 'contracts/sol6/**/*.sol'"
```
Comment (ilanDoron) [inline package.json]: please lint all solidity versions here.

**Review by ilanDoron — COMMENTED** [review]
First thing please make travis tests pass.
and see small comments.

**Review by ilanDoron — COMMENTED** [review]
we could also remove migrations.sol

**Review by ilanDoron — COMMENTED** [review]
note we have Token.sol both in sol5/mock and sol6/mock
no reason to keep both.

File: .solhint.json (hunk 15 above)
Comment (ayobuenavista) [reply .solhint.json]: Accidentally removed it when testing the linting. Returned now.

File: package.json (hunk 16 above)
Comment (ayobuenavista) [reply package.json]: Done.

File: README.md [hunk 17]
```diff
@@ -1,2 +1,33 @@
...
+
+## Setup
+1. Clone this repo
+2. `yarn`
+
+## Compilation
+1. `yarn compile` to compile contracts for all solidity versions.
+2. `yarn compileSol6` to compile only sol6 contracts
+
+## Testing with Buidler
```
Comment (ilanDoron) [inline README.md]: with buidler looks weird to me.
this is just testing.

File: .travis/report.sh [hunk 18]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+# node contractSizeReport.js $TRAVIS_PULL_REQUEST_BRANCH
```
Comment (ilanDoron) [inline .travis/report.sh]: I think this should run contract size report.
no?

File: .travis.yml [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+        - TEST_PART: Coverage
+
+install:
+  - yarn install
+
+before_script:
+  - yarn lint
+
+script:
+  - ./compile.sh
```
Comment (Anyhowclick) [inline .travis.yml]: Can consider changing to `yarn compile`

File: .travis/test-part.sh [hunk 20]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+# -*- firestarter: "shfmt -i 4 -ci -w %p" -*-
+
+set -euxo pipefail
+
+readonly test_part=${TEST_PART:-}
+
+case "$test_part" in
+Sol6)
+    npx buidler test --no-compile --config buidlerConfigSol6.js
```
Comment (Anyhowclick) [inline .travis/test-part.sh]: `npx` -> `yarn`

**Review by Anyhowclick — CHANGES_REQUESTED** [review]
Some small stuff

File: .travis.yml (hunk 19 above)
Comment (ayobuenavista) [reply .travis.yml]: Changed

File: .travis/test-part.sh (hunk 20 above)
Comment (ayobuenavista) [reply .travis/test-part.sh]: Changed

File: .travis.yml [hunk 21]
```diff
@@ -9,12 +9,12 @@ before_install:
 
 env:
   global:
-    - COVERAGE_BRANCH=master
+    - COVERAGE_BRANCH=reorg
```
Comment (ilanDoron) [inline .travis.yml]: so please remove all coverage stuff from travis

File: .travis.yml [hunk 22]
```diff
@@ -9,12 +9,12 @@ before_install:
...
 
 jobs:
   include:
     - env:
-        - TEST_PART: regression
+        - TEST_PART: Regression
     - env:
         - TEST_PART: Coverage
```
Comment (ilanDoron) [inline .travis.yml]: here as well...

File: .travis/test-part.sh [hunk 23]
```diff
@@ -6,14 +6,14 @@ set -euxo pipefail
 readonly test_part=${TEST_PART:-}
 
 case "$test_part" in
-Sol6)
-    npx buidler test --no-compile --config buidlerConfigSol6.js
+Regression)
+    yarn test
     ;;
 Coverage)
```
Comment (ilanDoron) [inline .travis/test-part.sh]: and this

File: buidlerConfigSol4.js [hunk 24]
```diff
@@ -149,7 +149,7 @@ module.exports = {
     tests: "./test/sol4"
   },
 
-  mocha: {
-    enableTimeouts: false
+  mocha: {
```
Comment (ilanDoron) [inline buidlerConfigSol4.js]: adding spaces

probably not required

File: buidlerConfigSol5.js [hunk 25]
```diff
@@ -149,7 +149,7 @@ module.exports = {
     tests: "./test/sol5"
   },
 
-  mocha: {
-    enableTimeouts: false
+  mocha: {	
+    enableTimeouts: false
```
Comment (ilanDoron) [inline buidlerConfigSol5.js]: adding spaces?

**Review by ilanDoron — COMMENTED** [review]
see comments

File: buidlerConfigSol4.js (hunk 24 above)
Comment (ilanDoron) [reply buidlerConfigSol4.js]: this comment is for last commit only.
here can't see the extra spaces.
just check spaces in editor.

File: package.json [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+    "c": "./compile.sh",
+    "compileSol6": "./compileSol6.sh",
+    "coverage": "./coverage.sh",
+    "lint": "yarn solhint 'contracts/**/*.sol' && yarn prettier -c './**/*.js'"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/KyberNetwork/kyber_utils_sc.git"
+  },
+  "author": "",
```
Comment (ilanDoron) [inline package.json]: maybe write 

Kyber SC team

File: package.json [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/KyberNetwork/kyber_utils_sc/issues"
+  },
+  "homepage": "https://github.com/KyberNetwork/kyber_utils_sc#readme",
+  "dependencies": {
+    "chai-as-promised": "7.1.1",
+    "compare-versions": "3.6.0",
```
Comment (ilanDoron) [inline package.json]: please make sure we use all packages.

I think we don't use ethers.
also not sure about compare-versions.
also truffle-assertions.
truffle-flattener
winston (its for fuzzer tests)

**Review by ducquangkstn — APPROVED** [review]

**Review by ilanDoron — APPROVED** [review]

**Review by Anyhowclick — APPROVED** [review]

**Pull request merged**